# **1. Import Library**

Pada tahap ini, Anda perlu mengimpor beberapa pustaka (library) Python yang dibutuhkan untuk analisis data dan pembangunan model machine learning.

In [ ]:
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
import matplotlib.pyplot as plt
from sklearn.model_selection import RandomizedSearchCV
import seaborn as sns
import geopandas as gpd
import numpy as np
import time
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score
from sklearn.metrics import confusion_matrix
from sklearn.linear_model import LogisticRegression
from google.colab import drive
from sklearn.metrics import classification_report
drive.mount('/content/drive')

Mounted at /content/drive


# **2. Memuat Dataset dari Hasil Clustering**

Memuat dataset hasil clustering dari file CSV ke dalam variabel DataFrame.

In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Dataset_Inisiasi.csv')
df

,Age,BMI,Systolic_BP,Diastolic_BP,Total_Cholesterol,HDL_Cholesterol,LDL_Cholesterol,Triglycerides,Smoking_Status,Physical_Activity_Level,Cluster
0,69,20.72,107.15,58.25,168.92,38.42,183.84,243.34,0,1,1
1,32,26.72,102.95,86.64,203.91,39.61,209.08,266.38,1,2,3
2,89,25.97,116.33,90.28,213.56,39.41,245.88,256.97,0,1,0
3,78,29.58,104.27,85.58,204.30,33.03,207.49,257.72,0,1,3
4,38,28.35,104.05,81.33,195.77,34.56,213.83,246.51,0,1,3
...,...,...,...,...,...,...,...,...,...,...,...
24995,87,15.98,129.52,93.01,177.30,40.23,199.20,229.35,0,1,1
24996,34,23.99,101.74,87.17,184.60,44.78,206.17,225.42,0,2,1
24997,27,26.73,101.74,87.02,166.87,41.85,221.07,224.64,0,1,1
24998,30,24.06,92.93,73.38,159.53,33.89,197.03,215.30,0,0,1


# **3. Data Splitting**

Tahap Data Splitting bertujuan untuk memisahkan dataset menjadi dua bagian: data latih (training set) dan data uji (test set).

In [ ]:
X = df.drop('Cluster', axis = 1)
y = df['Cluster']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

# **4. Membangun Model Klasifikasi**


## **a. Membangun Model Klasifikasi**

Setelah memilih algoritma klasifikasi yang sesuai, langkah selanjutnya adalah melatih model menggunakan data latih.

Berikut adalah rekomendasi tahapannya.
1. Pilih algoritma klasifikasi yang sesuai, seperti Logistic Regression, Decision Tree, Random Forest, atau K-Nearest Neighbors (KNN).
2. Latih model menggunakan data latih.

Algoritma RandomForest

In [ ]:
clf = RandomForestClassifier(random_state=42)
clf.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

Algoritma LogisticRegression

In [ ]:
clf_lin = LogisticRegression(random_state=42)
clf_lin.fit(X_train, y_train)

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=42)

Tulis narasi atau penjelasan algoritma yang Anda gunakan.

Algoritma yang akan saya gunakan adalah algoritma RandomForestClassifier. RandomForest adalah algoritma berbasis decision tree yang terdiri dari banyak pohon (ensemble learning). Algoritma ini bekerja dengan membuat banyak decision tree secara acak dan mengambil rata-rata prediksi dari semua pohon untuk menghasilkan hasil akhir yang lebih stabil dan akurat. Sebagai pembanding kita akan menggunakan algoritma logistic regression.

## **b. Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Lakukan prediksi menggunakan data uji.
2. Hitung metrik evaluasi seperti Accuracy dan F1-Score (Opsional: Precision dan Recall).
3. Buat confusion matrix untuk melihat detail prediksi benar dan salah.

RandomForest

Prediksi dengan menggunakan data tes/uji

In [ ]:
y_preds = clf.predict(X_test)

In [ ]:
print(f"Accurracy: {accuracy_score(y_test, y_preds)*100:.2f}%")
print(f"Precision: {precision_score(y_test, y_preds, average='weighted')}")
print(f"Recall: {recall_score(y_test, y_preds, average='weighted')}")
print(f"F1: {f1_score(y_test, y_preds, average='weighted')}")

Accurracy: 96.00%
Precision: 0.9600069960719629
Recall: 0.96
F1: 0.959992852057573


In [ ]:
confusion_matrix(y_test, y_preds)

pd.crosstab(y_test,y_preds,
            rownames = ["Actual Labels"],
            colnames = ["Predicted Labels"])

Predicted Labels,0,1,2,3
Actual Labels,,,,
0,1101,0,18,22
1,0,1066,17,18
2,30,20,1356,17
3,20,19,19,1277


LogisticRegression

In [ ]:
lin_y_preds = clf_lin.predict(X_test)

In [ ]:
print(f"Accurracy: {accuracy_score(y_test, lin_y_preds)*100:.2f}%")
print(f"Precision: {precision_score(y_test, lin_y_preds, average='weighted')}")
print(f"Recall: {recall_score(y_test, lin_y_preds, average='weighted')}")
print(f"F1: {f1_score(y_test, lin_y_preds, average='weighted')}")

Accurracy: 65.96%
Precision: 0.6575983572822218
Recall: 0.6596
F1: 0.6579876276726538


In [ ]:
confusion_matrix(y_test, lin_y_preds)

pd.crosstab(y_test,lin_y_preds,
            rownames = ["Actual Labels"],
            colnames = ["Predicted Labels"])

Predicted Labels,0,1,2,3
Actual Labels,,,,
0,631,25,210,275
1,25,665,224,187
2,182,209,1028,4
3,218,142,1,974


Tulis hasil evaluasi algoritma yang digunakan, jika Anda menggunakan 2 algoritma, maka bandingkan hasilnya.

**Random Forest**

Accurracy: 96.00%

Precision: 0.9600069960719629

Recall: 0.96

F1: 0.959992852057573


**Logistic Regression**

Accurracy: 65.96%

Precision: 0.6575983572822218

Recall: 0.6596

F1: 0.6579876276726538


Berdasarkan hasil evaluasi akurasi dan classification report, algoritma random forest memiliki performa yang sangat jauh lebih baik untuk dataset ini.  Random forest memiliki akurasi sebesar 96% dibandingkan dengan logistic regression yang hanya mampu mencapai 65.96%

## **c. Tuning Model Klasifikasi (Optional)**

Gunakan GridSearchCV, RandomizedSearchCV, atau metode lainnya untuk mencari kombinasi hyperparameter terbaik

In [ ]:
param_distributions = {"n_estimators": [10, 100, 200, 500, 1000, 1200],
                       "max_depth": [None, 5, 10, 20, 30],
                       "max_features": ["sqrt", "log2", None],
                       "min_samples_split": [2, 4, 6, 8],
                       "min_samples_leaf": [1, 2, 4, 8]}

Menggunakan RandomizedSearchCV untuk mencari hyperparameter terbaik

RandomForest

In [ ]:
start_time = time.time()

np.random.seed(42)

clf = RandomForestClassifier(n_jobs=-1)

n_iter = 30
rs_clf = RandomizedSearchCV(estimator=clf,
                            param_distributions=param_distributions,
                            n_iter=n_iter,
                            cv=5,
                            verbose=2)

rs_clf.fit(X_train, y_train);

end_time = time.time()
print(f"[INFO] Total time taken for {n_iter} random combinations of hyperparameters: {end_time - start_time:.2f} seconds.")

Fitting 5 folds for each of 30 candidates, totalling 150 fits
[CV] END max_depth=None, max_features=log2, min_samples_leaf=8, min_samples_split=2, n_estimators=10; total time=   0.3s
[CV] END max_depth=None, max_features=log2, min_samples_leaf=8, min_samples_split=2, n_estimators=10; total time=   0.3s
[CV] END max_depth=None, max_features=log2, min_samples_leaf=8, min_samples_split=2, n_estimators=10; total time=   0.3s
[CV] END max_depth=None, max_features=log2, min_samples_leaf=8, min_samples_split=2, n_estimators=10; total time=   0.3s
[CV] END max_depth=None, max_features=log2, min_samples_leaf=8, min_samples_split=2, n_estimators=10; total time=   0.5s
[CV] END max_depth=10, max_features=sqrt, min_samples_leaf=2, min_samples_split=2, n_estimators=1200; total time=  37.4s
[CV] END max_depth=10, max_features=sqrt, min_samples_leaf=2, min_samples_split=2, n_estimators=1200; total time=  36.5s
[CV] END max_depth=10, max_features=sqrt, min_samples_leaf=2, min_samples_split=2, n_estima

LogisticRegression

In [ ]:
start_time = time.time()

np.random.seed(42)

clf_lin = LogisticRegression(n_jobs=-1)

n_iter = 30
param_distributions_lin = {
    'C': np.logspace(-4, 4, 10),
    'penalty': ['l1', 'l2'],
    'solver': ['saga', 'liblinear'],
    'max_iter': [500, 1000, 2000]
}
rs_clf_lin = RandomizedSearchCV(estimator=clf_lin,
                            param_distributions=param_distributions_lin,
                            n_iter=n_iter,
                            cv=5,
                            verbose=2)

rs_clf_lin.fit(X_train, y_train);

end_time = time.time()

Fitting 5 folds for each of 30 candidates, totalling 150 fits
[CV] END C=0.046415888336127774, max_iter=2000, penalty=l1, solver=saga; total time=   8.3s
[CV] END C=0.046415888336127774, max_iter=2000, penalty=l1, solver=saga; total time=   7.0s
[CV] END C=0.046415888336127774, max_iter=2000, penalty=l1, solver=saga; total time=   9.8s
[CV] END C=0.046415888336127774, max_iter=2000, penalty=l1, solver=saga; total time=   7.0s
[CV] END C=0.046415888336127774, max_iter=2000, penalty=l1, solver=saga; total time=   8.7s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.046415888336127774, max_iter=2000, penalty=l2, solver=liblinear; total time=   0.4s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.046415888336127774, max_iter=2000, penalty=l2, solver=liblinear; total time=   0.4s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.046415888336127774, max_iter=2000, penalty=l2, solver=liblinear; total time=   0.4s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.046415888336127774, max_iter=2000, penalty=l2, solver=liblinear; total time=   0.4s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.046415888336127774, max_iter=2000, penalty=l2, solver=liblinear; total time=   0.4s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END ...C=0.0001, max_iter=1000, penalty=l1, solver=saga; total time=  11.7s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END ...C=0.0001, max_iter=1000, penalty=l1, solver=saga; total time=  11.8s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END ...C=0.0001, max_iter=1000, penalty=l1, solver=saga; total time=  11.9s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END ...C=0.0001, max_iter=1000, penalty=l1, solver=saga; total time=  11.7s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END ...C=0.0001, max_iter=1000, penalty=l1, solver=saga; total time=  11.6s
[CV] END C=0.3593813663804626, max_iter=1000, penalty=l2, solver=liblinear; total time=   0.5s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.3593813663804626, max_iter=1000, penalty=l2, solver=liblinear; total time=   0.6s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.3593813663804626, max_iter=1000, penalty=l2, solver=liblinear; total time=   0.5s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.3593813663804626, max_iter=1000, penalty=l2, solver=liblinear; total time=   0.5s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.3593813663804626, max_iter=1000, penalty=l2, solver=liblinear; total time=   0.5s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END C=0.005994842503189409, max_iter=500, penalty=l2, solver=saga; total time=   3.8s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END C=0.005994842503189409, max_iter=500, penalty=l2, solver=saga; total time=   4.6s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END C=0.005994842503189409, max_iter=500, penalty=l2, solver=saga; total time=   4.8s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END C=0.005994842503189409, max_iter=500, penalty=l2, solver=saga; total time=   3.8s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END C=0.005994842503189409, max_iter=500, penalty=l2, solver=saga; total time=   4.2s
[CV] END C=2.782559402207126, max_iter=1000, penalty=l1, solver=saga; total time=   9.2s
[CV] END C=2.782559402207126, max_iter=1000, penalty=l1, solver=saga; total time=  10.4s
[CV] END C=2.782559402207126, max_iter=1000, penalty=l1, solver=saga; total time=   9.8s
[CV] END C=2.782559402207126, max_iter=1000, penalty=l1, solver=saga; total time=   7.6s
[CV] END C=2.782559402207126, max_iter=1000, penalty=l1, solver=saga; total time=   9.7s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=21.54434690031882, max_iter=500, penalty=l1, solver=liblinear; total time=  24.9s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=21.54434690031882, max_iter=500, penalty=l1, solver=liblinear; total time=  23.0s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=21.54434690031882, max_iter=500, penalty=l1, solver=liblinear; total time=  23.2s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=21.54434690031882, max_iter=500, penalty=l1, solver=liblinear; total time=  23.9s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=21.54434690031882, max_iter=500, penalty=l1, solver=liblinear; total time=  22.3s
[CV] END ...C=0.0001, max_iter=2000, penalty=l2, solver=saga; total time=   0.8s
[CV] END ...C=0.0001, max_iter=2000, penalty=l2, solver=saga; total time=   0.9s
[CV] END ...C=0.0001, max_iter=2000, penalty=l2, solver=saga; total time=   0.6s
[CV] END ...C=0.0001, max_iter=2000, penalty=l2, solver=saga; total time=   0.5s
[CV] END ...C=0.0001, max_iter=2000, penalty=l2, solver=saga; total time=   0.5s
[CV] END C=0.046415888336127774, max_iter=1000, penalty=l1, solver=saga; total time=   6.6s
[CV] END C=0.046415888336127774, max_iter=1000, penalty=l1, solver=saga; total time=   8.8s
[CV] END C=0.046415888336127774, max_iter=1000, penalty=l1, solver=saga; total time=   9.6s
[CV] END C=0.046415888336127774, max_iter=1000, penalty=l1, solver=saga; total time=   7.0s
[CV] END C=0.046415888336127774, max_iter=1000, penalty=l1, solver=saga; total time=   9.4s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=1291.5496650148827, max_iter=2000, penalty=l2, solver=liblinear; total time=   0.6s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=1291.5496650148827, max_iter=2000, penalty=l2, solver=liblinear; total time=   0.7s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=1291.5496650148827, max_iter=2000, penalty=l2, solver=liblinear; total time=   0.6s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=1291.5496650148827, max_iter=2000, penalty=l2, solver=liblinear; total time=   0.6s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=1291.5496650148827, max_iter=2000, penalty=l2, solver=liblinear; total time=   0.6s
[CV] END C=0.000774263682681127, max_iter=1000, penalty=l2, solver=saga; total time=   1.8s
[CV] END C=0.000774263682681127, max_iter=1000, penalty=l2, solver=saga; total time=   1.8s
[CV] END C=0.000774263682681127, max_iter=1000, penalty=l2, solver=saga; total time=   2.5s
[CV] END C=0.000774263682681127, max_iter=1000, penalty=l2, solver=saga; total time=   2.5s
[CV] END C=0.000774263682681127, max_iter=1000, penalty=l2, solver=saga; total time=   1.8s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END C=2.782559402207126, max_iter=500, penalty=l2, solver=saga; total time=   3.9s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END C=2.782559402207126, max_iter=500, penalty=l2, solver=saga; total time=   4.3s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END C=2.782559402207126, max_iter=500, penalty=l2, solver=saga; total time=   5.0s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END C=2.782559402207126, max_iter=500, penalty=l2, solver=saga; total time=   3.9s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=2.782559402207126, max_iter=500, penalty=l2, solver=saga; total time=   4.0s
[CV] END C=0.0001, max_iter=2000, penalty=l2, solver=liblinear; total time=   0.3s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.0001, max_iter=2000, penalty=l2, solver=liblinear; total time=   0.3s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.0001, max_iter=2000, penalty=l2, solver=liblinear; total time=   0.3s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.0001, max_iter=2000, penalty=l2, solver=liblinear; total time=   0.3s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.0001, max_iter=2000, penalty=l2, solver=liblinear; total time=   0.3s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END C=0.046415888336127774, max_iter=500, penalty=l1, solver=saga; total time=   6.5s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END C=0.046415888336127774, max_iter=500, penalty=l1, solver=saga; total time=   5.6s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END C=0.046415888336127774, max_iter=500, penalty=l1, solver=saga; total time=   6.8s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END C=0.046415888336127774, max_iter=500, penalty=l1, solver=saga; total time=   5.3s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.046415888336127774, max_iter=500, penalty=l1, solver=saga; total time=   7.0s
[CV] END C=166.81005372000558, max_iter=1000, penalty=l1, solver=liblinear; total time=  24.0s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=166.81005372000558, max_iter=1000, penalty=l1, solver=liblinear; total time=  22.8s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=166.81005372000558, max_iter=1000, penalty=l1, solver=liblinear; total time=  25.5s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=166.81005372000558, max_iter=1000, penalty=l1, solver=liblinear; total time=  22.9s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=166.81005372000558, max_iter=1000, penalty=l1, solver=liblinear; total time=  25.0s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=166.81005372000558, max_iter=1000, penalty=l2, solver=liblinear; total time=   0.8s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=166.81005372000558, max_iter=1000, penalty=l2, solver=liblinear; total time=   0.6s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=166.81005372000558, max_iter=1000, penalty=l2, solver=liblinear; total time=   0.6s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=166.81005372000558, max_iter=1000, penalty=l2, solver=liblinear; total time=   0.6s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=166.81005372000558, max_iter=1000, penalty=l2, solver=liblinear; total time=   0.6s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=10000.0, max_iter=500, penalty=l1, solver=liblinear; total time=  25.3s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=10000.0, max_iter=500, penalty=l1, solver=liblinear; total time=  25.1s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=10000.0, max_iter=500, penalty=l1, solver=liblinear; total time=  24.7s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=10000.0, max_iter=500, penalty=l1, solver=liblinear; total time=  24.8s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=10000.0, max_iter=500, penalty=l1, solver=liblinear; total time=  24.9s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END ....C=0.0001, max_iter=500, penalty=l1, solver=saga; total time=   5.1s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END ....C=0.0001, max_iter=500, penalty=l1, solver=saga; total time=   6.7s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END ....C=0.0001, max_iter=500, penalty=l1, solver=saga; total time=   5.0s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END ....C=0.0001, max_iter=500, penalty=l1, solver=saga; total time=   6.6s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_sag.py:348: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  warnings.warn(


[CV] END ....C=0.0001, max_iter=500, penalty=l1, solver=saga; total time=   5.1s
[CV] END C=166.81005372000558, max_iter=1000, penalty=l1, solver=saga; total time=   9.9s
[CV] END C=166.81005372000558, max_iter=1000, penalty=l1, solver=saga; total time=  10.7s
[CV] END C=166.81005372000558, max_iter=1000, penalty=l1, solver=saga; total time=   8.9s
[CV] END C=166.81005372000558, max_iter=1000, penalty=l1, solver=saga; total time=  10.1s
[CV] END C=166.81005372000558, max_iter=1000, penalty=l1, solver=saga; total time=  10.3s
[CV] END C=1291.5496650148827, max_iter=2000, penalty=l1, solver=saga; total time=   8.8s
[CV] END C=1291.5496650148827, max_iter=2000, penalty=l1, solver=saga; total time=  10.5s
[CV] END C=1291.5496650148827, max_iter=2000, penalty=l1, solver=saga; total time=  10.3s
[CV] END C=1291.5496650148827, max_iter=2000, penalty=l1, solver=saga; total time=   9.7s
[CV] END C=1291.5496650148827, max_iter=2000, penalty=l1, solver=saga; total time=   8.5s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=2.782559402207126, max_iter=1000, penalty=l1, solver=liblinear; total time=  22.8s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=2.782559402207126, max_iter=1000, penalty=l1, solver=liblinear; total time=  24.4s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=2.782559402207126, max_iter=1000, penalty=l1, solver=liblinear; total time=  21.7s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=2.782559402207126, max_iter=1000, penalty=l1, solver=liblinear; total time=  24.8s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=2.782559402207126, max_iter=1000, penalty=l1, solver=liblinear; total time=  24.2s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.046415888336127774, max_iter=2000, penalty=l1, solver=liblinear; total time=   9.3s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.046415888336127774, max_iter=2000, penalty=l1, solver=liblinear; total time=   8.4s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.046415888336127774, max_iter=2000, penalty=l1, solver=liblinear; total time=   9.9s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.046415888336127774, max_iter=2000, penalty=l1, solver=liblinear; total time=   9.5s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.046415888336127774, max_iter=2000, penalty=l1, solver=liblinear; total time=   8.4s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.005994842503189409, max_iter=1000, penalty=l2, solver=liblinear; total time=   0.3s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.005994842503189409, max_iter=1000, penalty=l2, solver=liblinear; total time=   0.3s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.005994842503189409, max_iter=1000, penalty=l2, solver=liblinear; total time=   0.3s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.005994842503189409, max_iter=1000, penalty=l2, solver=liblinear; total time=   0.3s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.005994842503189409, max_iter=1000, penalty=l2, solver=liblinear; total time=   0.5s
[CV] END C=2.782559402207126, max_iter=2000, penalty=l2, solver=saga; total time=   6.8s
[CV] END C=2.782559402207126, max_iter=2000, penalty=l2, solver=saga; total time=   6.5s
[CV] END C=2.782559402207126, max_iter=2000, penalty=l2, solver=saga; total time=   7.3s
[CV] END C=2.782559402207126, max_iter=2000, penalty=l2, solver=saga; total time=   5.6s
[CV] END C=2.782559402207126, max_iter=2000, penalty=l2, solver=saga; total time=   7.2s
[CV] END C=0.046415888336127774, max_iter=1000, penalty=l2, solver=saga; total time=   5.3s
[CV] END C=0.046415888336127774, max_iter=1000, penalty=l2, solver=saga; total time=   7.5s
[CV] END C=0.046415888336127774, max_iter=1000, penalty=l2, solver=saga; total time=   5.6s
[CV] END C=0.046415888336127774, max_iter=1000, penalty=l2, solver=saga; total time=   6.8s
[CV] END C=0.046415888336127774, max_iter=1000, penalty=l2, solver=saga; total time=   5.4

/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.000774263682681127, max_iter=500, penalty=l2, solver=liblinear; total time=   0.2s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.000774263682681127, max_iter=500, penalty=l2, solver=liblinear; total time=   0.2s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.000774263682681127, max_iter=500, penalty=l2, solver=liblinear; total time=   0.2s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.000774263682681127, max_iter=500, penalty=l2, solver=liblinear; total time=   0.3s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=0.000774263682681127, max_iter=500, penalty=l2, solver=liblinear; total time=   0.3s
[CV] END ..C=10000.0, max_iter=1000, penalty=l2, solver=saga; total time=   7.0s
[CV] END ..C=10000.0, max_iter=1000, penalty=l2, solver=saga; total time=   6.3s
[CV] END ..C=10000.0, max_iter=1000, penalty=l2, solver=saga; total time=   7.3s
[CV] END ..C=10000.0, max_iter=1000, penalty=l2, solver=saga; total time=   5.5s
[CV] END ..C=10000.0, max_iter=1000, penalty=l2, solver=saga; total time=   7.3s
[CV] END C=21.54434690031882, max_iter=1000, penalty=l1, solver=saga; total time=   9.7s
[CV] END C=21.54434690031882, max_iter=1000, penalty=l1, solver=saga; total time=   8.9s
[CV] END C=21.54434690031882, max_iter=1000, penalty=l1, solver=saga; total time=  10.1s
[CV] END C=21.54434690031882, max_iter=1000, penalty=l1, solver=saga; total time=   9.6s
[CV] END C=21.54434690031882, max_iter=1000, penalty=l1, solver=saga; total time=   8.4s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=1291.5496650148827, max_iter=500, penalty=l1, solver=liblinear; total time=  25.2s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=1291.5496650148827, max_iter=500, penalty=l1, solver=liblinear; total time=  23.6s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=1291.5496650148827, max_iter=500, penalty=l1, solver=liblinear; total time=  23.0s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=1291.5496650148827, max_iter=500, penalty=l1, solver=liblinear; total time=  23.1s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


[CV] END C=1291.5496650148827, max_iter=500, penalty=l1, solver=liblinear; total time=  24.7s


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:1271: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 2.
  warnings.warn(


## **d. Evaluasi Model Klasifikasi setelah Tuning (Optional)**

Berikut adalah rekomendasi tahapannya.
1. Gunakan model dengan hyperparameter terbaik.
2. Hitung ulang metrik evaluasi untuk melihat apakah ada peningkatan performa.

RandomForest

In [ ]:
rs_clf.best_params_

{'n_estimators': 1000,
 'min_samples_split': 2,
 'min_samples_leaf': 1,
 'max_features': 'log2',
 'max_depth': 20}

In [ ]:
rs_y_preds = rs_clf.predict(X_test)
rs_y_preds

array([2, 2, 0, ..., 1, 1, 3])

In [ ]:
print(f"Accurracy: {accuracy_score(y_test, rs_y_preds)*100:.2f}%")
print(f"Precision: {precision_score(y_test, rs_y_preds, average='weighted')}")
print(f"Recall: {recall_score(y_test, rs_y_preds, average='weighted')}")
print(f"F1: {f1_score(y_test, rs_y_preds, average='weighted')}")

Accurracy: 96.10%
Precision: 0.9610602245331417
Recall: 0.961
F1: 0.9610057860280438


In [ ]:
confusion_matrix(y_test, rs_y_preds)

pd.crosstab(y_test,rs_y_preds,
            rownames = ["Actual Labels"],
            colnames = ["Predicted Labels"])

Predicted Labels,0,1,2,3
Actual Labels,,,,
0,1104,0,18,19
1,0,1058,17,26
2,30,15,1359,19
3,19,19,13,1284


LogisticRegression

In [ ]:
rs_clf_lin.best_params_

{'solver': 'liblinear',
 'penalty': 'l2',
 'max_iter': 2000,
 'C': np.float64(1291.5496650148827)}

In [ ]:
lin_rs_y_preds = rs_clf_lin.predict(X_test)

In [ ]:
print(f"Accurracy: {accuracy_score(y_test, lin_rs_y_preds)*100:.2f}%")
print(f"Precision: {precision_score(y_test, lin_rs_y_preds, average='weighted')}")
print(f"Recall: {recall_score(y_test, lin_rs_y_preds, average='weighted')}")
print(f"F1: {f1_score(y_test, lin_rs_y_preds, average='weighted')}")

Accurracy: 98.02%
Precision: 0.9803619941259705
Recall: 0.9802
F1: 0.9802053897795034


In [ ]:
confusion_matrix(y_test, lin_rs_y_preds)

pd.crosstab(y_test,lin_rs_y_preds,
            rownames = ["Actual Labels"],
            colnames = ["Predicted Labels"])

Predicted Labels,0,1,2,3
Actual Labels,,,,
0,1108,0,18,15
1,0,1072,17,12
2,7,8,1396,12
3,3,2,5,1325


## **e. Analisis Hasil Evaluasi Model Klasifikasi**

Berikut adalah **rekomendasi** tahapannya.
1. Bandingkan hasil evaluasi sebelum dan setelah tuning (jika dilakukan).
2. Identifikasi kelemahan model, seperti:
  - Precision atau Recall rendah untuk kelas tertentu.
  - Apakah model mengalami overfitting atau underfitting?
3. Berikan rekomendasi tindakan lanjutan, seperti mengumpulkan data tambahan atau mencoba algoritma lain jika hasil belum memuaskan.

## RandomForest

Training set

In [ ]:
y_train_preds = rs_clf.predict(X_train)

print("Metrics on the train set:")
print(f"Accuracy: {accuracy_score(y_train, y_train_preds) * 100:.2f}%")
print(f"Precision: {precision_score(y_train, y_train_preds, average='weighted'):.4f}")
print(f"Recall: {recall_score(y_train, y_train_preds, average='weighted'):.4f}")
print(f"F1-score: {f1_score(y_train, y_train_preds, average='weighted'):.4f}")


Metrics on the train set:
Accuracy: 100.00%
Precision: 1.0000
Recall: 1.0000
F1-score: 1.0000


Before Tuning

In [ ]:
print(f"Accurracy: {accuracy_score(y_test, y_preds)*100:.2f}%")
print(f"Precision: {precision_score(y_test, y_preds, average='weighted')}")
print(f"Recall: {recall_score(y_test, y_preds, average='weighted')}")
print(f"F1: {f1_score(y_test, y_preds, average='weighted')}")

Accurracy: 96.00%
Precision: 0.9600069960719629
Recall: 0.96
F1: 0.959992852057573


After Tuning

In [ ]:
print(f"Accurracy: {accuracy_score(y_test, rs_y_preds)*100:.2f}%")
print(f"Precision: {precision_score(y_test, rs_y_preds, average='weighted')}")
print(f"Recall: {recall_score(y_test, rs_y_preds, average='weighted')}")
print(f"F1: {f1_score(y_test, rs_y_preds, average='weighted')}")

Accurracy: 96.10%
Precision: 0.9610602245331417
Recall: 0.961
F1: 0.9610057860280438


Dari hasil tuning yang telah dilakukan, terjadi peningkatan yang baik pada accuracy, precision, recall, dan F1-score. Model sebelum tuning sudah terlihat sangat baik dengan akurasi 96%. Proses tuning hanya memberikan peningkatan sedikit ke 96.10%

In [ ]:
print(classification_report(y_test, rs_y_preds))

              precision    recall  f1-score   support

           0       0.96      0.97      0.96      1141
           1       0.97      0.96      0.96      1101
           2       0.97      0.96      0.96      1423
           3       0.95      0.96      0.96      1335

    accuracy                           0.96      5000
   macro avg       0.96      0.96      0.96      5000
weighted avg       0.96      0.96      0.96      5000



Model belajar terlalu baik di data Llatih (berpotensi overfitting)
Training Accuracy: 100% → Semua data latih diklasifikasikan dengan sempurna.

Accuracy meningkat dari 96.00% ➝ 96.10%

F1-score juga sedikit naik menunjukkan tuning RandomizedSearchCV berhasil menemukan kombinasi parameter yang lebih optimal, walaupun tidak drastis.

Semua kelas punya precision, recall, dan f1 hampir setara. Bisa diartikan model tidak bias ke salah satu kelas, dan semua cluster hasil clustering sebelumnya bisa dikenali dengan baik.

Hasil evaluasi model pada train set menunjukkan bahwa model mencapai 100% akurasi, yang lebih tinggi dibandingkan dengan 96% pada test set. Perbedaan yang kecil tidak mengindikasikan adanya overfitting

## LogisticRegression

Training

In [ ]:
lin_y_train_preds = rs_clf_lin.predict(X_train)

print("Metrics on the train set:")
print(f"Accuracy: {accuracy_score(y_train, lin_y_train_preds) * 100:.2f}%")
print(f"Precision: {precision_score(y_train, lin_y_train_preds, average='weighted'):.4f}")
print(f"Recall: {recall_score(y_train, lin_y_train_preds, average='weighted'):.4f}")
print(f"F1-score: {f1_score(y_train, lin_y_train_preds, average='weighted'):.4f}")

Metrics on the train set:
Accuracy: 97.84%
Precision: 0.9786
Recall: 0.9785
F1-score: 0.9785


Before Tuning

In [ ]:
print(f"Accurracy: {accuracy_score(y_test, lin_y_preds)*100:.2f}%")
print(f"Precision: {precision_score(y_test, lin_y_preds, average='weighted')}")
print(f"Recall: {recall_score(y_test, lin_y_preds, average='weighted')}")
print(f"F1: {f1_score(y_test, lin_y_preds, average='weighted')}")

Accurracy: 65.96%
Precision: 0.6575983572822218
Recall: 0.6596
F1: 0.6579876276726538


After Tuning

In [ ]:
print(f"Accurracy: {accuracy_score(y_test, lin_rs_y_preds)*100:.2f}%")
print(f"Precision: {precision_score(y_test, lin_rs_y_preds, average='weighted')}")
print(f"Recall: {recall_score(y_test, lin_rs_y_preds, average='weighted')}")
print(f"F1: {f1_score(y_test, lin_rs_y_preds, average='weighted')}")

Accurracy: 98.02%
Precision: 0.9803619941259705
Recall: 0.9802
F1: 0.9802053897795034


Tuning berhasil meningkatkan kinerja algoritma secara drastis

**Sebelum tuning:**

Accuracy: 65.96%

F1-score: 0.658

**Setelah tuning:**

Accuracy: 98.02%

F1-score: 0.980

In [ ]:
print(classification_report(y_test, lin_rs_y_preds))

              precision    recall  f1-score   support

           0       0.99      0.97      0.98      1141
           1       0.99      0.97      0.98      1101
           2       0.97      0.98      0.98      1423
           3       0.97      0.99      0.98      1335

    accuracy                           0.98      5000
   macro avg       0.98      0.98      0.98      5000
weighted avg       0.98      0.98      0.98      5000



- Sebelum tuning, algoritma logsitic regression sangat underperform.

- Setelah tuning, model menjadi sangat optimal.

- Tidak ada tanda overfitting besar karena train vs test score sangat dekat ( 97.8% vs 98.0%).

Setelah tuning, Logistic Regression mampu mengklasifikasikan hasil clustering dengan sangat baik (98% accuracy). Menunjukkan bahwa fitur-fitur seperti BMI, tekanan darah, kolesterol, dan aktivitas fisik memiliki pola yang cukup linear dalam memisahkan kelompok-kelompok gaya hidup dan kesehatan yang terbentuk dari clustering.

Berdasarkan hasil klasifikasi terhadap label yang diperoleh dari proses clustering, kedua algoritma yaitu random forest dan logistic regression menunjukkan performa yang sangat baik setelah dilakukan hyperparameter tuning. random forest mencapai akurasi sebesar 96.1%, sedangkan logistic regression bahkan lebih tinggi dengan akurasi 98.02%. Sebelumnya, logistic regression memiliki performa yang jauh lebih rendah, hanya mencapai 65.96%. Namun, model mengalami peningkatan signifikan setelah tuning yang menunjukkan pentingnya proses optimasi model.

Selain itu, kedua model menunjukkan distribusi performa yang seimbang di setiap label. Nilai precision, recall, dan f1-score kedua algoritma hampir merata. Setelah proses klasifikasi ini, bisa disimpulkan pemisahan antar label dapat dijelaskan dengan baik menggunakan hubungan linear, karena logistic regression yang bersifat linear mampu mengklasifikasikan dengan sangat akurat. Di sisi lain, random forest sebagai model non-linear tetap memberikan hasil yang stabil dan kompetitif, menunjukkan bahwa pola data juga mengandung sedikit kompleksitas. Secara keseluruhan, kedua model membuktikan bahwa hasil clustering yang dilakukan dapat diinterpretasikan dengan baik.